File: packages/schema/src/frame.ts
import { z } from "zod";

const bboxSchema = z.object({
  x: z.number(),
  y: z.number(),
  w: z.number().nonnegative(),
  h: z.number().nonnegative(),
});

export const ocrBubbleSchema = z.object({
  id: z.string().min(1),
  text: z.string(),
  bbox: bboxSchema,
  order: z.number().int().nonnegative(),
  confidence: z.number().min(0).max(1),
  language: z.enum(["ja", "zh", "en"]),
});

export const reviewedBubbleSchema = z.object({
  id: z.string().min(1),
  sourceBubbleId: z.string().min(1),
  textOriginal: z.string(),
  textEdited: z.string(),
  order: z.number().int().nonnegative(),
  kind: z.enum(["dialogue", "narration", "sfx", "ignore"]),
  speaker: z.string().min(1).optional(),
});

export const frameSchema = z.object({
  frameId: z.string().min(1),
  image: z.string().min(1),
  ocrFile: z.string().min(1),
  bubbles: z.array(ocrBubbleSchema),
  reviewedBubbles: z.array(reviewedBubbleSchema),
});

export type OcrBubble = z.infer<typeof ocrBubbleSchema>;
export type ReviewedBubble = z.infer<typeof reviewedBubbleSchema>;
export type Frame = z.infer<typeof frameSchema>;

File: apps/api/app/models/frame.py
from pydantic import BaseModel, ConfigDict, Field


class BoundingBox(BaseModel):
    x: float
    y: float
    w: float
    h: float


class OcrBubble(BaseModel):
    id: str
    text: str
    bbox: BoundingBox
    order: int
    confidence: float
    language: str


class ReviewedBubble(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: str
    source_bubble_id: str = Field(alias="sourceBubbleId")
    text_original: str = Field(alias="textOriginal")
    text_edited: str = Field(alias="textEdited")
    order: int
    kind: str
    speaker: str | None = None


class Frame(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    frame_id: str = Field(alias="frameId")
    image: str
    ocr_file: str = Field(alias="ocrFile")
    bubbles: list[OcrBubble]
    reviewed_bubbles: list[ReviewedBubble] = Field(alias="reviewedBubbles")
